PurchasesCoreSwift/Networking/Backend.swift

```diff
@@ -0,0 +1,650 @@
+
+        let cacheKey =
+        """
+        \(appUserID)-\(isRestore)-\(fetchToken)-\(productInfo?.cacheKey ?? "")
+        -\(offeringIdentifier ?? "")-\(observerMode)-\(subscriberAttributesByKey?.debugDescription ?? "")"
+        """
+
+        if add(callback: completion, key: cacheKey) {
+            return
+        }
+
+        if let productInfo = productInfo {
+            body.merge(productInfo.asDictionary()) { _, new in new }
+        }
+
+        if let subscriberAttributesByKey = subscriberAttributesByKey {
+
+            let attributesInBackendFormat = subscriberAttributesToDict(subscriberAttributes: subscriberAttributesByKey)
+            body["attributes"] = attributesInBackendFormat as AnyObject
+        }
+
+        if let offeringIdentifier = offeringIdentifier {
+            body["presented_offering_identifier"] = offeringIdentifier as NSString
+        }
+
+        httpClient.performPOSTRequest(serially: true,
+                                      path: "/receipts",
+                                      requestBody: body,
+                                      headers: authHeaders) { statusCode, response, error in
+            let callbacks = self.getPurchaserInfoCallbacksAndClearCache(forKey: cacheKey)
+            for callback in callbacks {
+                self.handle(purchaserInfoResponse: response, statusCode: statusCode, error: error, completion: callback)
+            }
+        }
+    }
+
+    @objc
+    public func getSubscriberData(appUserID: String, completion: @escaping BackendPurchaserInfoResponseHandler) {
+        guard let appUserID = try? escapedAppUserID(appUserID: appUserID) else {
+            completion(nil, ErrorUtils.missingAppUserIDError())
+            return
+        }
+
+        let path = "/subscribers/\(appUserID)"
+
+        if add(callback: completion, key: path) {
+            return
+        }
+
+        httpClient.performGETRequest(serially: true, path: path, headers: authHeaders) { statusCode, response, error in
+            for completion in self.getPurchaserInfoCallbacksAndClearCache(forKey: path) {
+                print("Hi")
+                self.handle(purchaserInfoResponse: response,
+                            statusCode: statusCode,
+                            error: error,
+                            completion: completion)
+            }
+        }
+    }
+
+    // swiftlint:disable function_parameter_count
+    @objc(postOfferForSigning:withProductIdentifier:subscriptionGroup:receiptData:appUserID:completion:)
+    public func post(offerIdForSigning offerIdentifier: String,
+                     productIdentifier: String,
+                     subscriptionGroup: String,
+                     receiptData: Data,
+                     appUserID: String,
+                     completion: @escaping OfferSigningResponseHandler) {
+    // swiftlint:enable function_parameter_count
+        let fetchToken = receiptData.base64EncodedString(options: .lineLength64Characters)
```

Comment:
perhaps we could extract this logic since we have it in a couple of places?

Reply:
💯